File: Darts/src/burlton/dartzee/code/screen/reporting/ReportingResultsScreen.java
package burlton.dartzee.code.screen.reporting;
import burlton.dartzee.code.bean.ScrollTableDartsGame;
import burlton.dartzee.code.reporting.ReportParameters;
import burlton.dartzee.code.reporting.ReportResultWrapper;
import burlton.dartzee.code.reporting.ReportingSqlUtil;
import burlton.dartzee.code.screen.EmbeddedScreen;
import burlton.dartzee.code.screen.ScreenCache;
import burlton.desktopcore.code.util.TableUtil;

import javax.swing.*;
import java.awt.*;
import java.awt.event.ActionEvent;
import java.sql.Timestamp;
import java.util.ArrayList;

public class ReportingResultsScreen extends EmbeddedScreen
{
	private ReportParameters rp = null;
	private ArrayList<Object[]> cachedRows = null;
	
	public ReportingResultsScreen() 
	{
		add(tableResults, BorderLayout.CENTER);
		
		JPanel panel = new JPanel();
		add(panel, BorderLayout.NORTH);
		
		
		panel.add(btnConfigureColumns);
		btnConfigureColumns.addActionListener(this);
	}
	
	private final JButton btnConfigureColumns = new JButton("Configure Columns...");
	private final ScrollTableDartsGame tableResults = new ScrollTableDartsGame();

	@Override
	public String getScreenName()
	{
		return "Game Report";
	}

	@Override
	public void initialise()
	{
		buildTable(true);
	}

	private void buildTable(boolean runSql)
	{
		TableUtil.DefaultModel model = new TableUtil.DefaultModel();
		model.addColumn("Game");
		model.addColumn("Type");
		model.addColumn("Players");
		model.addColumn("Start Date");
		model.addColumn("Finish Date");
		model.addColumn("Match");
		
		if (runSql)
		{
			ArrayList<ReportResultWrapper> wrappers = ReportingSqlUtil.runReport(rp);
			cachedRows = ReportResultWrapper.getTableRowsFromWrappers(wrappers);
		}
		
		for (Object[] row : cachedRows)
		{
			model.addRow(row);
		}
		
		tableResults.setRowName("game");
		tableResults.setModel(model);
		tableResults.setColumnWidths("60;160;-1;DT;DT;100");
		tableResults.sortBy(0, false);
		
		setRenderersAndComparators();
		stripOutRemovedColumns();
	}
	private void setRenderersAndComparators()
	{
		tableResults.setRenderer(3, TableUtil.TIMESTAMP_RENDERER);
		tableResults.setRenderer(4, TableUtil.TIMESTAMP_RENDERER);
		
		tableResults.setComparator(3, (Timestamp t1, Timestamp t2) -> t1.compareTo(t2));
		tableResults.setComparator(3, (Timestamp t1, Timestamp t2) -> t1.compareTo(t2));
	}
	private void stripOutRemovedColumns()
	{
		ConfigureReportColumnsDialog dlg = ScreenCache.getConfigureReportColumnsDialog();

		int columns = tableResults.getColumnCount();
		for (int i=columns-1; i>=0; i--)
		{
			String columnName = tableResults.getColumnName(i);
			if (!dlg.includeColumn(columnName))
			{
				tableResults.removeColumn(i);
			}
		}
	}
	
	public void setReportParameters(ReportParameters rp)
	{
		this.rp = rp;
	}
	
	@Override
	public EmbeddedScreen getBackTarget()
	{
		return ScreenCache.getScreen(ReportingSetupScreen.class);
	}
	
	@Override
	public void actionPerformed(ActionEvent arg0)
	{
		if (arg0.getSource() == btnConfigureColumns)
		{
			ConfigureReportColumnsDialog dlg = ScreenCache.getConfigureReportColumnsDialog();
			dlg.setLocationRelativeTo(this);
			dlg.setVisible(true);
			
			buildTable(false);
		}
		else
		{
			super.actionPerformed(arg0);
		}
	}	
}

File: src/main/kotlin/dartzee/screen/reporting/ReportingResultsScreen.kt
package dartzee.screen.reporting

import dartzee.bean.ScrollTableDartsGame
import dartzee.core.util.TableUtil
import dartzee.reporting.ReportParameters
import dartzee.reporting.ReportResultWrapper
import dartzee.reporting.runReport
import dartzee.screen.EmbeddedScreen
import dartzee.screen.ScreenCache
import java.awt.BorderLayout
import java.awt.event.ActionEvent
import java.sql.Timestamp
import javax.swing.JButton
import javax.swing.JPanel

class ReportingResultsScreen(
    private val configureColumnsDialog: ConfigureReportColumnsDialog =
        ConfigureReportColumnsDialog()
) : EmbeddedScreen() {
    var rp: ReportParameters? = null
    private var cachedRows = emptyList<Array<Any>>()

    private val btnConfigureColumns = JButton("Configure Columns...")
    private val tableResults = ScrollTableDartsGame()

    init {
        add(tableResults, BorderLayout.CENTER)

        val panel = JPanel()
        add(panel, BorderLayout.NORTH)

        panel.add(btnConfigureColumns)
        btnConfigureColumns.addActionListener(this)
    }

    override fun getScreenName() = "Game Report"

    override fun initialise() {
        buildTable(true)
    }

    private fun buildTable(runSql: Boolean) {
        val model = TableUtil.DefaultModel()
        model.addColumn("Game")
        model.addColumn("Type")
        model.addColumn("Players")
        model.addColumn("Start Date")
        model.addColumn("Finish Date")
        model.addColumn("Match")

        if (runSql) {
            val wrappers = runReport(rp)
            cachedRows = ReportResultWrapper.getTableRowsFromWrappers(wrappers)
        }

        model.addRows(cachedRows)

        tableResults.setRowName("game")
        tableResults.model = model
        tableResults.setColumnWidths("60;160;-1;DT;DT;100")
        tableResults.sortBy(0, false)

        setRenderersAndComparators()
        stripOutRemovedColumns()
    }

    private fun setRenderersAndComparators() {
        tableResults.setRenderer(3, TableUtil.TimestampRenderer())
        tableResults.setRenderer(4, TableUtil.TimestampRenderer())

        tableResults.setComparator(3, compareBy<Timestamp> { it })
        tableResults.setComparator(4, compareBy<Timestamp> { it })
    }

    private fun stripOutRemovedColumns() {
        val columns = tableResults.columnCount
        for (i in columns - 1 downTo 0) {
            val columnName = tableResults.getColumnName(i)
            if (configureColumnsDialog.excludedColumns().contains(columnName)) {
                tableResults.removeColumn(i)
            }
        }
    }

    override fun getBackTarget() = ScreenCache.get<ReportingSetupScreen>()

    override fun actionPerformed(arg0: ActionEvent) {
        when (arg0.source) {
            btnConfigureColumns -> {
                configureColumnsDialog.setLocationRelativeTo(this)
                configureColumnsDialog.isVisible = true

                buildTable(false)
            }
            else -> super.actionPerformed(arg0)
        }
    }
}
